### PR TITLE
[Java] Alignment checks and optimizations.

### PR DESCRIPTION
**Pull request by vyazelenko**
This PR contains the following changes:
- Changes `UnsafeBuffer#verifyAlignment` to throw if the underlying buffer is a `byte[]`/`HeapByteBuffer` and the `AtomicBuffer#STRICT_ALIGNMENT_CHECKS == true` (which is always `true` on non-x86 platforms and defaults to `false` for x86). There is also a new property `agrona.strict.alignment.checks` that can be used to configure `AtomicBuffer#STRICT_ALIGNMENT_CHECKS` on x86.
- Adds a common base class `AbstractMutableDirectBuffer` for all buffers.
- Performs bounds checks in all buffers only if the `DirectBuffer#SHOULD_BOUNDS_CHECK` is `true`.
- Optimizes `equals/hashCode/compareTo` methods to perform comparisons in 8-byte chunks.
- Optimizes `setMemory` for small length (i.e. `< 100`).
  * **`JDK 1.8.0_352-b08 (Zulu 8.66.0.15-CA-linux64)`**:
    * Before:
    ```
    Benchmark                     (buffer)  (index)  (length)  Mode  Cnt   Score   Error  Units
    SetMemoryBenchmark.benchmark     ARRAY        0         1  avgt   30  14.649 ± 2.198  ns/op
    SetMemoryBenchmark.benchmark     ARRAY        0         2  avgt   30  14.008 ± 2.368  ns/op
    SetMemoryBenchmark.benchmark     ARRAY        0         4  avgt   30  17.128 ± 2.412  ns/op
    SetMemoryBenchmark.benchmark     ARRAY        0        16  avgt   30  12.835 ± 1.415  ns/op
    SetMemoryBenchmark.benchmark     ARRAY        0        50  avgt   30  15.416 ± 1.813  ns/op
    SetMemoryBenchmark.benchmark     ARRAY        0        99  avgt   30  20.767 ± 2.361  ns/op
    SetMemoryBenchmark.benchmark     ARRAY        0       128  avgt   30  14.977 ± 2.603  ns/op
    SetMemoryBenchmark.benchmark     ARRAY        0      1024  avgt   30  23.627 ± 2.266  ns/op
    SetMemoryBenchmark.benchmark     ARRAY        3         1  avgt   30  13.290 ± 1.508  ns/op
    SetMemoryBenchmark.benchmark     ARRAY        3         2  avgt   30  16.798 ± 2.479  ns/op
    SetMemoryBenchmark.benchmark     ARRAY        3         4  avgt   30  13.835 ± 2.291  ns/op
    SetMemoryBenchmark.benchmark     ARRAY        3        16  avgt   30  17.777 ± 2.343  ns/op
    SetMemoryBenchmark.benchmark     ARRAY        3        50  avgt   30  18.266 ± 2.564  ns/op
    SetMemoryBenchmark.benchmark     ARRAY        3        99  avgt   30  14.311 ± 2.651  ns/op
    SetMemoryBenchmark.benchmark     ARRAY        3       128  avgt   30  16.778 ± 3.095  ns/op
    SetMemoryBenchmark.benchmark     ARRAY        3      1024  avgt   30  22.541 ± 2.581  ns/op
    SetMemoryBenchmark.benchmark    DIRECT        0         1  avgt   30  14.341 ± 1.477  ns/op
    SetMemoryBenchmark.benchmark    DIRECT        0         2  avgt   30  14.537 ± 1.824  ns/op
    SetMemoryBenchmark.benchmark    DIRECT        0         4  avgt   30  13.382 ± 1.574  ns/op
    SetMemoryBenchmark.benchmark    DIRECT        0        16  avgt   30  13.814 ± 1.684  ns/op
    SetMemoryBenchmark.benchmark    DIRECT        0        50  avgt   30  14.311 ± 1.536  ns/op
    SetMemoryBenchmark.benchmark    DIRECT        0        99  avgt   30  16.438 ± 2.899  ns/op
    SetMemoryBenchmark.benchmark    DIRECT        0       128  avgt   30  15.499 ± 2.601  ns/op
    SetMemoryBenchmark.benchmark    DIRECT        0      1024  avgt   30  17.603 ± 1.391  ns/op
    SetMemoryBenchmark.benchmark    DIRECT        3         1  avgt   30  14.227 ± 1.307  ns/op
    SetMemoryBenchmark.benchmark    DIRECT        3         2  avgt   30  12.735 ± 1.240  ns/op
    SetMemoryBenchmark.benchmark    DIRECT        3         4  avgt   30  14.369 ± 1.912  ns/op
    SetMemoryBenchmark.benchmark    DIRECT        3        16  avgt   30  12.997 ± 1.497  ns/op
    SetMemoryBenchmark.benchmark    DIRECT        3        50  avgt   30  16.562 ± 1.825  ns/op
    SetMemoryBenchmark.benchmark    DIRECT        3        99  avgt   30  15.144 ± 2.709  ns/op
    SetMemoryBenchmark.benchmark    DIRECT        3       128  avgt   30  14.106 ± 2.206  ns/op
    SetMemoryBenchmark.benchmark    DIRECT        3      1024  avgt   30  17.582 ± 1.686  ns/op
    ```
    * After:
    ```
    Benchmark                     (buffer)  (index)  (length)  Mode  Cnt   Score   Error  Units
    SetMemoryBenchmark.benchmark     ARRAY        0         1  avgt   30   3.955 ± 0.684  ns/op
    SetMemoryBenchmark.benchmark     ARRAY        0         2  avgt   30   2.941 ± 0.981  ns/op
    SetMemoryBenchmark.benchmark     ARRAY        0         4  avgt   30   2.536 ± 0.669  ns/op
    SetMemoryBenchmark.benchmark     ARRAY        0        16  avgt   30   2.123 ± 0.756  ns/op
    SetMemoryBenchmark.benchmark     ARRAY        0        50  avgt   30   4.165 ± 0.603  ns/op
    SetMemoryBenchmark.benchmark     ARRAY        0        99  avgt   30   5.964 ± 0.009  ns/op
    SetMemoryBenchmark.benchmark     ARRAY        0       128  avgt   30  13.759 ± 2.352  ns/op
    SetMemoryBenchmark.benchmark     ARRAY        0      1024  avgt   30  22.077 ± 2.825  ns/op
    SetMemoryBenchmark.benchmark     ARRAY        3         1  avgt   30   4.093 ± 0.729  ns/op
    SetMemoryBenchmark.benchmark     ARRAY        3         2  avgt   30   3.594 ± 1.009  ns/op
    SetMemoryBenchmark.benchmark     ARRAY        3         4  avgt   30   2.147 ± 0.002  ns/op
    SetMemoryBenchmark.benchmark     ARRAY        3        16  avgt   30   3.106 ± 1.031  ns/op
    SetMemoryBenchmark.benchmark     ARRAY        3        50  avgt   30   3.879 ± 0.456  ns/op
    SetMemoryBenchmark.benchmark     ARRAY        3        99  avgt   30   5.963 ± 0.005  ns/op
    SetMemoryBenchmark.benchmark     ARRAY        3       128  avgt   30  17.162 ± 3.180  ns/op
    SetMemoryBenchmark.benchmark     ARRAY        3      1024  avgt   30  21.454 ± 2.806  ns/op
    SetMemoryBenchmark.benchmark    DIRECT        0         1  avgt   30   3.040 ± 0.825  ns/op
    SetMemoryBenchmark.benchmark    DIRECT        0         2  avgt   30   2.447 ± 0.885  ns/op
    SetMemoryBenchmark.benchmark    DIRECT        0         4  avgt   30   2.784 ± 0.779  ns/op
    SetMemoryBenchmark.benchmark    DIRECT        0        16  avgt   30   2.900 ± 0.989  ns/op
    SetMemoryBenchmark.benchmark    DIRECT        0        50  avgt   30   3.767 ± 0.497  ns/op
    SetMemoryBenchmark.benchmark    DIRECT        0        99  avgt   30   5.958 ± 0.004  ns/op
    SetMemoryBenchmark.benchmark    DIRECT        0       128  avgt   30  12.587 ± 1.180  ns/op
    SetMemoryBenchmark.benchmark    DIRECT        0      1024  avgt   30  16.979 ± 1.453  ns/op
    SetMemoryBenchmark.benchmark    DIRECT        3         1  avgt   30   2.592 ± 0.641  ns/op
    SetMemoryBenchmark.benchmark    DIRECT        3         2  avgt   30   1.817 ± 0.418  ns/op
    SetMemoryBenchmark.benchmark    DIRECT        3         4  avgt   30   3.065 ± 0.881  ns/op
    SetMemoryBenchmark.benchmark    DIRECT        3        16  avgt   30   2.559 ± 0.941  ns/op
    SetMemoryBenchmark.benchmark    DIRECT        3        50  avgt   30   4.352 ± 0.615  ns/op
    SetMemoryBenchmark.benchmark    DIRECT        3        99  avgt   30   5.963 ± 0.008  ns/op
    SetMemoryBenchmark.benchmark    DIRECT        3       128  avgt   30  12.572 ± 0.346  ns/op
    SetMemoryBenchmark.benchmark    DIRECT        3      1024  avgt   30  19.653 ± 1.878  ns/op
    ```
  * **`JDK 17.0.5+8-LTS (Zulu17.38+21-CA)`**:
    * Before
    ```
    Benchmark                     (buffer)  (index)  (length)  Mode  Cnt   Score   Error  Units 
    SetMemoryBenchmark.benchmark     ARRAY        0         1  avgt   30  12.494 ± 2.055  ns/op 
    SetMemoryBenchmark.benchmark     ARRAY        0         2  avgt   30  19.167 ± 3.801  ns/op 
    SetMemoryBenchmark.benchmark     ARRAY        0         4  avgt   30  12.657 ± 3.123  ns/op 
    SetMemoryBenchmark.benchmark     ARRAY        0        16  avgt   30  15.464 ± 3.782  ns/op 
    SetMemoryBenchmark.benchmark     ARRAY        0        50  avgt   30  17.246 ± 4.214  ns/op 
    SetMemoryBenchmark.benchmark     ARRAY        0        99  avgt   30  11.735 ± 1.566  ns/op 
    SetMemoryBenchmark.benchmark     ARRAY        0       128  avgt   30  18.543 ± 3.839  ns/op 
    SetMemoryBenchmark.benchmark     ARRAY        0      1024  avgt   30  43.998 ± 2.139  ns/op 
    SetMemoryBenchmark.benchmark     ARRAY        3         1  avgt   30  14.853 ± 3.427  ns/op 
    SetMemoryBenchmark.benchmark     ARRAY        3         2  avgt   30  18.022 ± 4.022  ns/op 
    SetMemoryBenchmark.benchmark     ARRAY        3         4  avgt   30  14.156 ± 3.396  ns/op 
    SetMemoryBenchmark.benchmark     ARRAY        3        16  avgt   30  16.115 ± 4.140  ns/op 
    SetMemoryBenchmark.benchmark     ARRAY        3        50  avgt   30  13.352 ± 2.445  ns/op 
    SetMemoryBenchmark.benchmark     ARRAY        3        99  avgt   30  14.967 ± 3.696  ns/op 
    SetMemoryBenchmark.benchmark     ARRAY        3       128  avgt   30  15.635 ± 2.853  ns/op 
    SetMemoryBenchmark.benchmark     ARRAY        3      1024  avgt   30  43.398 ± 1.878  ns/op 
    SetMemoryBenchmark.benchmark    DIRECT        0         1  avgt   30  13.349 ± 2.735  ns/op 
    SetMemoryBenchmark.benchmark    DIRECT        0         2  avgt   30  12.829 ± 3.334  ns/op 
    SetMemoryBenchmark.benchmark    DIRECT        0         4  avgt   30  14.884 ± 3.716  ns/op 
    SetMemoryBenchmark.benchmark    DIRECT        0        16  avgt   30  11.411 ± 2.723  ns/op 
    SetMemoryBenchmark.benchmark    DIRECT        0        50  avgt   30  18.107 ± 3.446  ns/op 
    SetMemoryBenchmark.benchmark    DIRECT        0        99  avgt   30  16.096 ± 3.472  ns/op 
    SetMemoryBenchmark.benchmark    DIRECT        0       128  avgt   30  13.879 ± 2.218  ns/op 
    SetMemoryBenchmark.benchmark    DIRECT        0      1024  avgt   30  43.442 ± 1.872  ns/op 
    SetMemoryBenchmark.benchmark    DIRECT        3         1  avgt   30  11.484 ± 2.000  ns/op 
    SetMemoryBenchmark.benchmark    DIRECT        3         2  avgt   30  11.102 ± 2.536  ns/op 
    SetMemoryBenchmark.benchmark    DIRECT        3         4  avgt   30  10.455 ± 1.989  ns/op 
    SetMemoryBenchmark.benchmark    DIRECT        3        16  avgt   30  17.389 ± 3.247  ns/op 
    SetMemoryBenchmark.benchmark    DIRECT        3        50  avgt   30  15.959 ± 3.418  ns/op 
    SetMemoryBenchmark.benchmark    DIRECT        3        99  avgt   30  15.110 ± 3.683  ns/op 
    SetMemoryBenchmark.benchmark    DIRECT        3       128  avgt   30  17.741 ± 3.418  ns/op 
    SetMemoryBenchmark.benchmark    DIRECT        3      1024  avgt   30  43.613 ± 1.863  ns/op 
    ```
    * After
    ```
    Benchmark                     (buffer)  (index)  (length)  Mode  Cnt   Score   Error  Units
    SetMemoryBenchmark.benchmark     ARRAY        0         1  avgt   30   2.735 ± 0.764  ns/op
    SetMemoryBenchmark.benchmark     ARRAY        0         2  avgt   30   3.627 ± 0.896  ns/op
    SetMemoryBenchmark.benchmark     ARRAY        0         4  avgt   30   3.252 ± 0.488  ns/op
    SetMemoryBenchmark.benchmark     ARRAY        0        16  avgt   30   3.840 ± 0.872  ns/op
    SetMemoryBenchmark.benchmark     ARRAY        0        50  avgt   30   5.052 ± 0.451  ns/op
    SetMemoryBenchmark.benchmark     ARRAY        0        99  avgt   30   6.891 ± 0.314  ns/op
    SetMemoryBenchmark.benchmark     ARRAY        0       128  avgt   30  16.197 ± 3.030  ns/op
    SetMemoryBenchmark.benchmark     ARRAY        0      1024  avgt   30  44.708 ± 1.908  ns/op
    SetMemoryBenchmark.benchmark     ARRAY        3         1  avgt   30   3.091 ± 0.923  ns/op
    SetMemoryBenchmark.benchmark     ARRAY        3         2  avgt   30   2.841 ± 0.790  ns/op
    SetMemoryBenchmark.benchmark     ARRAY        3         4  avgt   30   3.667 ± 0.636  ns/op
    SetMemoryBenchmark.benchmark     ARRAY        3        16  avgt   30   3.306 ± 0.903  ns/op
    SetMemoryBenchmark.benchmark     ARRAY        3        50  avgt   30   5.207 ± 0.464  ns/op
    SetMemoryBenchmark.benchmark     ARRAY        3        99  avgt   30   6.933 ± 0.306  ns/op
    SetMemoryBenchmark.benchmark     ARRAY        3       128  avgt   30  14.091 ± 1.937  ns/op
    SetMemoryBenchmark.benchmark     ARRAY        3      1024  avgt   30  43.036 ± 1.984  ns/op
    SetMemoryBenchmark.benchmark    DIRECT        0         1  avgt   30   3.144 ± 0.890  ns/op
    SetMemoryBenchmark.benchmark    DIRECT        0         2  avgt   30   3.303 ± 0.937  ns/op
    SetMemoryBenchmark.benchmark    DIRECT        0         4  avgt   30   2.862 ± 0.004  ns/op
    SetMemoryBenchmark.benchmark    DIRECT        0        16  avgt   30   3.149 ± 0.805  ns/op
    SetMemoryBenchmark.benchmark    DIRECT        0        50  avgt   30   4.737 ± 0.304  ns/op
    SetMemoryBenchmark.benchmark    DIRECT        0        99  avgt   30   6.076 ± 0.359  ns/op
    SetMemoryBenchmark.benchmark    DIRECT        0       128  avgt   30  13.046 ± 0.862  ns/op
    SetMemoryBenchmark.benchmark    DIRECT        0      1024  avgt   30  41.710 ± 1.965  ns/op
    SetMemoryBenchmark.benchmark    DIRECT        3         1  avgt   30   2.778 ± 0.792  ns/op
    SetMemoryBenchmark.benchmark    DIRECT        3         2  avgt   30   2.827 ± 0.786  ns/op
    SetMemoryBenchmark.benchmark    DIRECT        3         4  avgt   30   3.681 ± 0.491  ns/op
    SetMemoryBenchmark.benchmark    DIRECT        3        16  avgt   30   3.621 ± 0.830  ns/op
    SetMemoryBenchmark.benchmark    DIRECT        3        50  avgt   30   5.018 ± 0.409  ns/op
    SetMemoryBenchmark.benchmark    DIRECT        3        99  avgt   30   6.499 ± 0.305  ns/op
    SetMemoryBenchmark.benchmark    DIRECT        3       128  avgt   30  12.195 ± 0.068  ns/op
    SetMemoryBenchmark.benchmark    DIRECT        3      1024  avgt   30  42.864 ± 2.153  ns/op
    ```
- `ExpandableDirectByteBuffer` now aligns the underlying `DirectByteBuffer` to 32-bytes.
- `BufferAlignmentAgent` checks were optimized and refactored.

### CODE REVIEW

File: agrona/src/main/java/org/agrona/concurrent/AtomicBuffer.java [hunk 1]
```diff
@@ -16,16 +16,47 @@
...
+     * {@link #STRICT_ALIGNMENT_CHECKS_PROP_NAME} system property.
+     *
+     * @see AtomicBuffer#STRICT_ALIGNMENT_CHECKS_PROP_NAME
+     */
+    boolean STRICT_ALIGNMENT_CHECKS = "true".equals(
+        SystemUtil.getProperty(STRICT_ALIGNMENT_CHECKS_PROP_NAME, SystemUtil.isX86Arch() ? "false" : "true"));
+
     /**
      * Verify that the underlying buffer is correctly aligned to prevent word tearing and other ordering issues.
+     * <p>
```
Comment (mikeb01) [inline agrona/src/main/java/org/agrona/concurrent/AtomicBuffer.java:54]: I think we need to go into a lot more detail here about the issues that can occur if accesses are unaligned, including the issues we have seen on Arm.  It should serve as a point of reference.  We should `@see` to the put/get methods and mention that indexes need to be aligned as well.

Comment (vyazelenko) [reply agrona/src/main/java/org/agrona/concurrent/AtomicBuffer.java:54]: Good point. I'll extend the JavaDoc here.

File: agrona/src/main/java/org/agrona/concurrent/AtomicBuffer.java [hunk 2]
```diff
@@ -16,16 +16,47 @@
...
+     */
+    String STRICT_ALIGNMENT_CHECKS_PROP_NAME = "agrona.strict.alignment.checks";
+
+    /**
+     * Should alignment checks for atomic operations be done or not. Controlled by the
+     * {@link #STRICT_ALIGNMENT_CHECKS_PROP_NAME} system property.
+     *
+     * @see AtomicBuffer#STRICT_ALIGNMENT_CHECKS_PROP_NAME
+     */
+    boolean STRICT_ALIGNMENT_CHECKS = "true".equals(
```
[the remark this answers is not in the document]
Comment (vyazelenko) [reply agrona/src/main/java/org/agrona/concurrent/AtomicBuffer.java:49]: See discussion in Slack.

File: agrona/src/main/java/org/agrona/concurrent/AtomicBuffer.java (hunk 1 above)
Comment (vyazelenko) [reply agrona/src/main/java/org/agrona/concurrent/AtomicBuffer.java:60]: Done.

**Review by mjpt777 — APPROVED** [review]
Nice work.

**Pull request merged**